i.eodag: add operators to queryables and allow multiple values

HamedElgizery opened the pull request:
This PR fixes the current faulty query option, and add the ability to use **operators** (eq, lt, gt, le, etc...). 
It also allows the query option to have multiple queryables (emulating the **AND** relation), and allows each queryable to have multiple value (emulating the **OR** relation).

Querying section is also added to the manual, with examples on how to use the query option.
Multiple use cases for i.eodag is also added to the manual.

HamedElgizery committed:
Fix query option

HamedElgizery committed:
Allow queryable to have multiple values

HamedElgizery committed:
Add strip to trim strings

HamedElgizery committed:
Check query option validity before searching to save time if there is an error

veroandreo commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -548,16 +550,14 @@ def filter_result(search_result, geometry, **kwargs):
+    start_date = None
+    end_date = None
 
     # If neither a geometry is provided as a parameter
     # nor a vector map is provided through "options",
     # then none of the geometry filtering will take place.
```
Why not? The region should be considered and the geometry filtering should happen with the region if no map is provided. Region can be as big or as small as user wants, so why not using it the same way map is used? Doesn't get_aoi fallback to get_bb if no vector is provided?

HamedElgizery replied:
> Doesn't get_aoi fallback to get_bb if no vector is provided?

Right, yes it does. I will fix that.

veroandreo commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -580,13 +580,84 @@ def filter_result(search_result, geometry, **kwargs):
             geometry=geometry, minimum_overlap=int(minimum_overlap)
         )
 
-    if cloud_cover:
-        search_result = search_result.filter_property(
-            operator="le", cloudCover=int(cloud_cover)
-        )
```
Why do you remove this here?

HamedElgizery replied:
Because if we need to filter by clouds, we can say:
`query="cloudCover=40|lt"`  (clouds < 40)
`query="cloudCover=40|gt"`  (clouds > 40)

And so it becomes more flexible, operators wise.

Also because now the code differentiate between the queryables and the normal explicitly provided options. So `clouds` option is used for searching, while `query="cloudCover={value}|{operator}` is used for filtering.


I am trying to distinguish between searching and filtering here, because there are some cases where we do already have the products beforehand, e.g. saved in a geojson file, and we just want to filter them out... In this case and similar ones, only the `query` option will be used for filtering these products, and the other provided options will be ignored, because they are only intended to be used when searching and not filtering. What do you think?

veroandreo replied:
For me, this adds a layer of complexity to the most common use case. The other modules and some other tools around use clouds to pass just an integer that means less than that. I wouldn't remove that simplicity which coincides with the most common usage. 

Still, I do not fully understand how search and filter modes will work in the context of i.eodag. Will users need to run the module twice or everything will happen in the background (desired)?

HamedElgizery replied:
> For me, this adds a layer of complexity to the most common use case. The other modules and some other tools around use clouds to pass just an interest that means less than that. I wouldn't remove that simplicity which coincides with the most common usage.

Okay, makes sense, I will put it back.
> Still, I do not fully understand how search and filter modes will work in the context of i.eodag. Will users need to run the module twice or everything will happen in the background (desired)?

No the user doesn't need to run the module twice, it happens in the background...

Here is a pseudo diagram of the code flow:
[searching from scratch, with options] -> [filtering] -> [output results]
[read ids from Text file] -> [search with ids] -> [filtering] -> [output results]
[read products from Geojson file] -> [filtering] -> [output results]

So filtering always takes place at the end before outputting the results.

veroandreo replied:
Thanks for the explanation, @HamedElgizery. Please make sure to explain all the module's logic in the manual page and add relevant examples.

HamedElgizery replied:
On it. I am currently working on the manual, specifically adding examples of how to use the query options and what it is capable of. I will push the manual changes when I reach a first reasonable draft.

HamedElgizery committed:
Add query_parser function and refactor code

HamedElgizery committed:
Restrict queryables to queryables with types str, int, float, or Literal

HamedElgizery committed:
Add querying section with examples to manual

echoix commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -788,13 +910,17 @@ def print_eodag_queryables(**kwargs):
         potential_type = info.__args__[0]
         if potential_type.__name__ != "Optional":
-            assert potential_type.__name__ in types_options
+            assert potential_type.__name__ in METADATA_TYPES
```
assert in Python code like this doesn't do what you think it does. It is in fact useless, at it is not guaranteed to be executed. It can be disabled, or not present in some installations. If the condition checked is required to be checked for correct execution of the logic, it should be written as such, like using an if.
Many Python linters will try to warn about this.


You can see here 
https://docs.python.org/3.12/reference/simple_stmts.html#the-assert-statement that this syntax
`assert_stmt ::=  "assert" expression ["," expression]`
, ie `assert 1==2, "that expression should be false"`

Is equivalent to
```python
if __debug__:
    if not expression1: raise AssertionError(expression2)
```

ie: 

```python
if __debug__:
    if not 1==2:
        raise AssertionError("that expression should be false")
```


You can also read

https://dbader.org/blog/catching-bogus-python-asserts


Or the ruff rule S101 https://docs.astral.sh/ruff/rules/assert/

HamedElgizery replied:
I didn't know that about the assert statements. Thanks for the references, @echoix!
I will fix this problem.

HamedElgizery committed:
Change assert statements to if statements

HamedElgizery committed:
Add explanation of some use cases to manual

ninsbl commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -25,15 +25,250 @@ <h3>WARNING: I.EODAG IS UNDER DEVELOPMENT. THIS IS AN EXPERIMENTAL VERSION.</h3>
+acquired, from the EODAG YAML config file.  User have to specify the config
+file path location through the <b>config</b> option, otherwise <em>i.eodag</em>
+will use the credentials found in the default config file
+<em>~/.config/eodag/eodag.yml</em>
```
Storing passwords in plain text files is a security issue. It should at least be possible to fetch them from env-variables. But that is for another PR...

veroandreo replied:
it is possible apparently: https://eodag.readthedocs.io/en/stable/notebooks/api_user_guide/1_overview.html#Configure

veroandreo committed:
i.eodag manual: language and style fixes, sections re-ordered

veroandreo commented:
@HamedElgizery I extensively modified the manual page to demonstrate the usual writing style. Please, have a look to check that I got things right (i.e., search products intersecting the computational region should be the default). I removed the -e flag in one example, as that flag is no longer there. Also, I believe some other options might be explained in the description section. 

And, as far as I recall from eodag itself, intersects is the default, so `area_relation` should have `answer:Intersects` as default.  

I leave the code review to @ninsbl and @lucadelu :)

lucadelu commented:
I looked at the code and I don't see big problems. I tested the pull request and most of the command I tried it works.

I have a problem trying to query using the id, when I use the id parameter it return an empty result. It could be a eodag problem and not related to the GRASS implementation...

```
GRASS utm32n/lucadelu:i.eodag > python i.eodag.py -l producttype=S2_MSI_L2A query="start=2022-05-25T00:00:00,end=2022-06-01T23:59:00"
WARNING: Experimental Version...
WARNING: This module is still under development, and its behaviour is not
         guaranteed to be reliable
3 scenes(s) found.
S2B_MSIL2A_20220526T101559_N0400_R065_T32TPS_20220526T134056 2022-05-26T10:15:59 50% S2MSI2A
S2A_MSIL2A_20220531T101611_N0400_R065_T32TPS_20220531T163910 2022-05-31T10:16:11 59% S2MSI2A
S2A_MSIL2A_20220528T100601_N0400_R022_T32TPS_20220528T162819 2022-05-28T10:06:01 70% S2MSI2A
GRASS utm32n/lucadelu:i.eodag > python i.eodag.py -l producttype=S2_MSI_L2A query="id=S2B_MSIL2A_20220526T101559_N0400_R065_T32TPS_20220526T134056\eq"
WARNING: Experimental Version...
WARNING: This module is still under development, and its behaviour is not
         guaranteed to be reliable
0 scenes(s) found.
```

neteler commented:
I made an initial test, great work!

A few observations:

Error to be caught when **input vector map not present** (just try with a non-existing one):

```python
GRASS latlong_wgs84/testarea:~ > i.eodag -l start=2017-10-21 end=2017-10-21 producttype=S2_MSI_L2A provider=cop_dataspace map=non_existing_bbox_map
WARNING: Experimental Version...
WARNING: This module is still under development, and its behaviour is not
         guaranteed to be reliable
ERROR: Vector map <non_existing_bbox_map> not found
Traceback (most recent call last):
  File "/home/mneteler/software/grass84/dist.x86_64-pc-linux-gnu/scripts/i.eodag", line 1144, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/mneteler/software/grass84/dist.x86_64-pc-linux-gnu/scripts/i.eodag", line 1057, in main
    geometry = get_aoi(options["map"])
               ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mneteler/software/grass84/dist.x86_64-pc-linux-gnu/scripts/i.eodag", line 296, in get_aoi
    if gs.vector_info_topo(vector)["areas"] <= 0:
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mneteler/software/grass84/dist.x86_64-pc-linux-gnu/etc/python/grass/script/vector.py", line 182, in vector_info_topo
    s = read_command("v.info", flags="t", layer=layer, map=map, env=env)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mneteler/software/grass84/dist.x86_64-pc-linux-gnu/etc/python/grass/script/core.py", line 554, in read_command
    return handle_errors(returncode, stdout, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mneteler/software/grass84/dist.x86_64-pc-linux-gnu/etc/python/grass/script/core.py", line 366, in handle_errors
    raise CalledModuleError(module=module, code=code, returncode=returncode)
grass.exceptions.CalledModuleError: Module run `v.info -t layer=1 map=non_existing_bbox_map` ended with an error.
The subprocess ended with a non-zero return code: 1. See errors above the traceback or in the error output.
```

And it would be good to have a **default cloud value** defined and visible to the user as it is unclear from the description, to which threshold `cloud` is set per default (could be 20% or whatever):

```
clouds   Maximum cloud cover percentage for scene [0, 100]
default: 20
```

And: which is the **default `output` directory?**

`output   Name for output directory where to store downloaded scenes data`

Indeed I found the downloaded S2 data in `$HOME/tmp/grass8-mneteler-619940/` (Linux) which is unexpected for me. Say, `r.out.gdal` etc. use the current directory as the default which is probably more intuitive. 

Thanks again for your work, my first `i.eodag` downloads are running :-)

neteler commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -2,38 +2,272 @@ <h2>DESCRIPTION</h2>
+<em>i.eodag</em> reads user credentials from the EODAG YAML config file.  
+Users have to specify the config file path through the <b>config</b> 
+option, otherwise <em>i.eodag</em> will use the credentials found in 
+the default config file <em>~/.config/eodag/eodag.yml</em>
```
```suggestion
the default config file <em>~/.config/eodag/eodag.yml</em> which is auto-generated the
first time eodag is used after an install in a local directory (`$HOME/.config/eodag/eodag.yml`
on Linux and BSD; XXXX on Windows; YYYY on Mac).
```

I suggest to add a note where the config is... (dunno where it would be on Windows or Mac).

Yet it looks like a chicken-egg problem: `pip install eodag` will install the software but what will generate the config? At first run on `i.eodag` if we do not want to suggest to run the `eodag` CLI tool? Perhaps I overlook something.

veroandreo replied:
The installation of eodag creates the file in the default location, then the user can modify it there or create a different one and place it somewhere else.

HamedElgizery replied:
> I suggest to add a note where the config is... (dunno where it would be on Windows or Mac).

I tested it on Mac, and it would be in the same location  `$HOME/.config/eodag`. Don't know about Windows though, yet.

> Yet it looks like a chicken-egg problem: `pip install eodag` will install the software but what will generate the config? At first run on `i.eodag` if we do not want to suggest to run the `eodag` CLI tool? Perhaps I overlook something.

Maybe we could add a flag `i`, for initialization, to use when running i.eodag for the first time. Just for the sake of making it less confusing, and then maybe we could use that flag to write a custom GRASS GIS EODAG config file, if we need to.

veroandreo replied:
Can't that just happen within i.eodag (with a check or so) without another flag? Users should be able to create their config files or modify the default one. 

At this point, we need to move forward with the other objectives of the GSoC project.

HamedElgizery committed:
Add <p> tags to seperate titles

HamedElgizery committed:
Set default area_relation to Intersects, and default clouds to 20

HamedElgizery committed:
Handle non-existent vector map

HamedElgizery commented:
> @HamedElgizery I extensively modified the manual page to demonstrate the usual writing style. Please, have a look to check that I got things right (i.e., search products intersecting the computational region should be the default). I removed the -e flag in one example, as that flag is no longer there. Also, I believe some other options might be explained in the description section.

Thanks, @veroandreo. It is more consistent and clear now. I have gone through it and everything looks right. I only added some `<p>` tags to separate the subtitles from the content in this commit dfb364a.
 
> And, as far as I recall from eodag itself, intersects is the default, so `area_relation` should have `answer:Intersects` as default.

Just added that.

HamedElgizery commented:
> I looked at the code and I don't see big problems. I tested the pull request and most of the command I tried it works.
> 
> I have a problem trying to query using the id, when I use the id parameter it return an empty result. It could be a eodag problem and not related to the GRASS implementation...
> 
> ```
> GRASS utm32n/lucadelu:i.eodag > python i.eodag.py -l producttype=S2_MSI_L2A query="start=2022-05-25T00:00:00,end=2022-06-01T23:59:00"
> WARNING: Experimental Version...
> WARNING: This module is still under development, and its behaviour is not
>          guaranteed to be reliable
> 3 scenes(s) found.
> S2B_MSIL2A_20220526T101559_N0400_R065_T32TPS_20220526T134056 2022-05-26T10:15:59 50% S2MSI2A
> S2A_MSIL2A_20220531T101611_N0400_R065_T32TPS_20220531T163910 2022-05-31T10:16:11 59% S2MSI2A
> S2A_MSIL2A_20220528T100601_N0400_R022_T32TPS_20220528T162819 2022-05-28T10:06:01 70% S2MSI2A
> GRASS utm32n/lucadelu:i.eodag > python i.eodag.py -l producttype=S2_MSI_L2A query="id=S2B_MSIL2A_20220526T101559_N0400_R065_T32TPS_20220526T134056\eq"
> WARNING: Experimental Version...
> WARNING: This module is still under development, and its behaviour is not
>          guaranteed to be reliable
> 0 scenes(s) found.
> ```

To filter by the id you need to use the queryable `title` instead.
So giving that you are using the same AOI, the following should work:
```
python i.eodag.py -l producttype=S2_MSI_L2A query="title=S2B_MSIL2A_20220526T101559_N0400_R065_T32TPS_20220526T134056\eq"
```
Or...
You can use search by ID from scratch as:
```
python i.eodag.py -l producttype=S2_MSI_L2A id=S2B_MSIL2A_20220526T101559_N0400_R065_T32TPS_20220526T134056
```

HamedElgizery committed:
Fix filtering of products that shouldn't be filter when using the id
option

Faulty filtering would take place without user permission, when searching with ids.
This commit fixes that.

HamedElgizery committed:
Set clouds option default value to 100

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -530,7 +531,91 @@ def dates_to_iso_format():
+    return query_dicts
+
+
+def filter_result(search_result, geometry=None, queryables=None, **kwargs):
```
Given that all product properties can be used for searching, they should probably be used in the search stage already to minimize search time... But you can leave that for another day, as I guess there are som caveats e.g. when searching for data publised in a given time-frame...

HamedElgizery replied:
If you are talking about the queryables...
The searching parameters is inconsistent among the providers, see below:
https://eodag.readthedocs.io/en/stable/add_provider.html#opensearch-parameters-csv

So filtering, after done with searching, makes things less error prone... and AFAIK queryables are intended for filtering the data with the EODAG crunch after the search.

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -585,12 +671,50 @@ def filter_result(search_result, geometry, **kwargs):
+
+    # Remove duplictes that might be created while filtering
     search_result = remove_duplicates(search_result)
+    if start_date or end_date:
+        search_result = search_result.filter_date(start=start_date, end=end_date)
```
Is there a point of filtering by date a second time?

HamedElgizery replied:
When you are searching for product normally, there is no need to filter a second time.

Yet, filtering a second time here is necessary in the case that you have already done the searching and saved the search result into a GeoJSON file, and then later you want to filter the GeoJSON file according to different parameters, including the date. Thus, the second filtering is necessary to handle that case (because it would actually be the first time filtering by date, for that GeoJSON file).

veroandreo replied:
Yeah, this is how it works with eodag, first search (which will probably be enough in most cases) and then filter within search results. 

Within GRASS, the filtering could also be done with SQL (in the GUI or with v.db.select or db.* tools) if the user saves the search results as footprints. The attr table of the footprint vector map should contain all queryables.

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -530,7 +531,91 @@ def dates_to_iso_format():
+        for value in map(str.strip, values.split("|")):
+            if value == "":
+                continue
+            if value.find("\\") != -1:
```
The `\` is a bit awkward choice as it is used for masking and may make it more difficult to handle on command line. Another singlecharacter separator like `|`, or `#`, or `;` would be probably a bit more appropriate....

HamedElgizery replied:
I will change it to `;`.

ninsbl commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -25,15 +25,250 @@ <h3>WARNING: I.EODAG IS UNDER DEVELOPMENT. THIS IS AN EXPERIMENTAL VERSION.</h3>
+Note that the <b>print</b> option only gives a subset of the avaible queryables,
+and you can in fact use any of the product's properties, for filtering.
+If you are not sure about the all the available properties,
+you can run a generic search with the <b>j</b> flag and setting <b>limit=1</b>,
```
Maybe better to implement that as convenience for the user instead of using the `list_queryables` function that can be missleading in terms of search options...

HamedElgizery replied:
Agree, I will create another PR with that, probably after getting done with i.landsat.download, if it is not urgent.

ninsbl reviewed (approved):
I do have some comments, but they are not urgent. So, I think we can merge.
As mentioned, tests would be nice, ideally based on the examples you added...

HamedElgizery committed:
Change \ to ; in query

HamedElgizery committed:
Update i.eodag.html

veroandreo commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -2,38 +2,272 @@ <h2>DESCRIPTION</h2>
+To be able to download data through <em>i.eodag</em>, users will
+need to register for the providers of interest.
+<em>i.eodag</em> reads user credentials from the EODAG YAML config file.
+Users have to specify the config file path through the <b>config</b>
+option, otherwise <em>i.eodag</em> will use the credentials found in
+the default config file <em>~/.config/eodag/eodag.yml</em> which is auto-generated the
+first time eodag is used after an install in `$HOME/.config/eodag/eodag.yml`.
```
```suggestion
Users have to specify the configuration file path through the <b>config</b>
option, otherwise <em>i.eodag</em> will use the credentials found in
the default config file which is auto-generated the first time EODAG is used 
after installation. The configuration file is stored by default in 
<code>$HOME/.config/eodag/eodag.yml</code>.
```

veroandreo committed:
Update src/imagery/i.eodag/i.eodag.html

veroandreo commented:
I will merge this one once tests finish. The remaining "questions" were moved here: https://github.com/HamedElgizery/grass-addons/issues/17

HamedElgizery commented:
> > And, as far as I recall from eodag itself, intersects is the default, so `area_relation` should have `answer:Intersects` as default.
> 
> Just added that.

I reverted this change (setting area_relation to Intersects as a default value), and now it has no default value. The reason is that the default value for area_relation makes i.eodag do unintended filtering for products passed from a GeoJSON file. If filtering of products from a GeoJSON file with respect to the area_relation is needed, then area_relation needs to be specified explicitly.

It was reverted here 44150b7569cbd8381b49660b97b512de73358581.
